### PR TITLE
whitespace: add option check-postblock-prekeyword

### PR DESCRIPTION
**Pull request by yubaoquan**
check-postblock-prekeyword

#### Example:

before lint:
```
try {
}catch {
}finally {
}

if () {
}else {
}
```
after lint:
```
try {
} catch {
} finally {
}


if () {
} else {
}
```

#### PR checklist

- [x] Addresses an existing issue: #746
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [x] Documentation update

#### CHANGELOG.md entry:

[new-rule] `check-postblock-prekeyword`
[new-rule]

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/tslint](https://github.com/palantir/tslint), @yubaoquan! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

**Comment by yubaoquan** [comment]
signed the individual CLA on October 12, 2017 at 12:03 AM. @palantirtech

File: src/rules/whitespaceRule.ts [hunk 1]
```diff
@@ -50,14 +51,17 @@ export class Rule extends Lint.Rules.AbstractRule {
             * \`"check-type"\` checks for whitespace before a variable type specification.
             * \`"check-typecast"\` checks for whitespace between a typecast and its target.
             * \`"check-type-operator"\` checks for whitespace between type operators \`|\` and \`&\`.
-            * \`"check-preblock"\` checks for whitespace before the opening brace of a block`,
+            * \`"check-preblock"\` checks for whitespace before the opening brace of a block.
+            * \`"check-postblock-prekeyword"\` checks for whitespace between the closing brace of a block
+            and the following keyword (\`} else\` / \`} catch\` / \`} while\`).`,
```
Comment (ajafff) [inline src/rules/whitespaceRule.ts]: what's the difference to `one-line` with `"check-whitespace"`?

File: src/rules/whitespaceRule.ts [hunk 2]
```diff
@@ -122,6 +128,21 @@ function walk(ctx: Lint.WalkContext<Options>) {
...
+                    const nextToken = utils.getNextToken(node as ts.Block)!;
+                    if (!utils.isSameLine(sourceFile, node.end, nextToken.getStart())) {
+                        break;
+                    }
+                    switch (nextToken.kind) {
+                        case ts.SyntaxKind.WhileKeyword:
+                        case ts.SyntaxKind.ElseKeyword:
+                        case ts.SyntaxKind.CatchKeyword:
+                        case ts.SyntaxKind.FinallyKeyword:
+                            if (node.end === nextToken.getStart()) {
```
Comment (ajafff) [inline src/rules/whitespaceRule.ts]: you can simply use `checkForTrailingWhitespace(node.end)`

File: src/rules/whitespaceRule.ts [hunk 3]
```diff
@@ -122,6 +128,21 @@ function walk(ctx: Lint.WalkContext<Options>) {
                 if (options.preblock) {
                     checkForTrailingWhitespace(node.getFullStart());
                 }
+                if (options.postblockPrekeyword) {
+                    const nextToken = utils.getNextToken(node as ts.Block)!;
+                    if (!utils.isSameLine(sourceFile, node.end, nextToken.getStart())) {
+                        break;
+                    }
+                    switch (nextToken.kind) {
+                        case ts.SyntaxKind.WhileKeyword:
```
Comment (ajafff) [inline src/rules/whitespaceRule.ts]: This may do a bit too much. That `WhileKeyword` could be from the `while (...) {}` statement following the block.
That may not be a big deal, because it will only complain about such code:
```ts
if (condition) { /* code */ }while(condition) {};
                             ~ [missing whitespace]
```

Anyway, there's an easier way to do this check without `getNextToken`:

```ts
const parent = node.parent!;
if (isDoStatement(parent) || // Block is followed by WhiteKeyword
    isIfStatement(parent) && parent.elseStatement !== undefined && parent.thenStatement === node || // Block is followed by ElseKeyword
    isTryStatement(parent) && (
        parent.tryBlock === node || // Block is followed by CatchKeyword or FinallyKeyword
        parent.catchClause !== undefined && parent.catchClause.block === node // Block is followed by FinallyKeyword
    )
) {
    // do stuff
}
```
Well, while writing it down I realized it's not really easier. But it's clearer what kind of statement is checked.

File: src/rules/whitespaceRule.ts (hunk 1 above)
Comment (yubaoquan) [reply src/rules/whitespaceRule.ts]: I reviewed the `one-line` rule code and its test case. Didn't find that rule solve the situation of this PR. And I tried to add `one-line` with all its options to my test file, the linter didn't report error of missing whitespace between closing bracket and `else`

File: src/rules/whitespaceRule.ts (hunk 3 above)
Comment (yubaoquan) [reply src/rules/whitespaceRule.ts]: OK. I wrote a function to check conditions as you mentions.

File: src/rules/whitespaceRule.ts (hunk 2 above)
Comment (yubaoquan) [reply src/rules/whitespaceRule.ts]: Changed as you said.

File: src/rules/whitespaceRule.ts (hunk 1 above)
Comment (ajafff) [reply src/rules/whitespaceRule.ts]: #3200 includes the check for whitespace before `else`.
The only thing missing is whitespace before `while` in `do {...} while (...);` loops. But that could easily be added.

Comment (yubaoquan) [reply src/rules/whitespaceRule.ts]: So is my pr going to be merged?

File: src/rules/whitespaceRule.ts [hunk 4]
```diff
@@ -306,7 +320,16 @@ function walk(ctx: Lint.WalkContext<Options>) {
...
     });
-
+    function matchPostblockPreKeyword(node: ts.Node): boolean {
+        const parent = node.parent!;
+        const isDoWhile = utils.isDoStatement(parent);
+        const isIfElse = utils.isIfStatement(parent)
+            && parent.elseStatement !== undefined
+            && parent.elseStatement !== node;
+        const isTryCatchFinally = utils.isTryStatement(parent)
+            && (parent.tryBlock === node || parent.catchClause === node);
```
Comment (ajafff) [inline src/rules/whitespaceRule.ts]: You can use the following to avoid the additional `case` for `CatchClause` above (sorry my initial suggestion got that part wrong):
```ts
const isTryCatchFinally = utils.isTryStatement(parent) && parent.tryBlock === node ||
    utils.isCatchClause(parent) && parent.parent!.finallyBlock !== undefined;
```

File: src/rules/whitespaceRule.ts [hunk 5]
```diff
@@ -306,7 +320,16 @@ function walk(ctx: Lint.WalkContext<Options>) {
                 }
         }
     });
-
+    function matchPostblockPreKeyword(node: ts.Node): boolean {
```
Comment (ajafff) [inline src/rules/whitespaceRule.ts]: Consider moving this function out of the `walk` closure. It doesn't use any variables from that scope, so it doesn't need to be here.

File: src/rules/whitespaceRule.ts (hunk 1 above)
Comment (ajafff) [reply src/rules/whitespaceRule.ts]: I have no preference into which rule the functionality should go. However, we generally try to avoid duplication.
I'll leave that decision to @adidahiya

File: src/rules/whitespaceRule.ts (hunk 4 above)
Comment (yubaoquan) [reply src/rules/whitespaceRule.ts]: Done

File: src/rules/whitespaceRule.ts (hunk 5 above)
Comment (yubaoquan) [reply src/rules/whitespaceRule.ts]: Done

File: src/rules/whitespaceRule.ts (hunk 1 above)
Comment (adidahiya) [reply src/rules/whitespaceRule.ts]: @yubaoquan do you generally use `one-line` with `"check-whitespace"` in your lint config? if so, I think we should just add the additional check @ajafff suggested to `one-line` instead of a new option here.

**Comment by JKillian** [comment]
it's a little confusing, but functionality should go in the `one-line` rule likely, see https://github.com/palantir/tslint/pull/3308#discussion_r153646487
